diff -ru lua-4.0/src/llex.c lua-4.0-block-comment/src/llex.c
--- lua-4.0/src/llex.c	Fri Oct 20 18:39:03 2000
+++ lua-4.0-block-comment/src/llex.c	Fri Dec 14 12:19:09 2001
@@ -195,7 +195,10 @@
     switch (LS->current) {
       case EOZ:
         save(L, '\0', l);
-        luaX_error(LS, "unfinished long string", TK_STRING);
+        if (seminfo)
+          luaX_error(LS, "unfinished long string", TK_STRING);
+        else
+          luaX_error(LS, "unfinished comment", TK_EOS);
         break;  /* to avoid warnings */
       case '[':
         save_and_next(L, LS, l);
@@ -217,12 +220,15 @@
         inclinenumber(LS);
         continue;
       default:
-        save_and_next(L, LS, l);
+        if (seminfo)	/* no need to save complete comment */
+          save(L, LS->current, l);
+        next(LS);
     }
   } endloop:
   save_and_next(L, LS, l);  /* skip the second ']' */
   save(L, '\0', l);
-  seminfo->ts = luaS_newlstr(L, L->Mbuffer+2, l-5);
+  if (seminfo)
+    seminfo->ts = luaS_newlstr(L, L->Mbuffer+2, l-5);
 }
 
 
@@ -297,7 +303,11 @@
       case '-':
         next(LS);
         if (LS->current != '-') return '-';
-        do { next(LS); } while (LS->current != '\n' && LS->current != EOZ);
+        if (next(LS) == '[' && next(LS) == '[')
+          read_long_string(LS, NULL);
+        else
+          while (LS->current != '\n' && LS->current != EOZ)
+            next(LS);
         continue;
 
       case '[':
